File: labs/memory_bound/huge_pages_1/GenerateMesh.cpp
#include "GenerateMesh.hpp"

#include <algorithm>
#include <numeric>

auto generateMesh(unsigned n_nodes_x, unsigned n_nodes_y, double *x, double *y,
                  unsigned seed) -> std::vector<std::array<unsigned, 2>> {
  const unsigned n_nodes = n_nodes_x * n_nodes_y;

  // Topology
  std::vector<std::array<unsigned, 2>> topology;
  topology.reserve((n_nodes_x - 1) * (n_nodes_y - 1) * 3 + (n_nodes_x - 1) +
                   (n_nodes_y - 1));
  for (unsigned j = 0; j < n_nodes_y - 1; ++j) {
    for (unsigned i = 0; i < n_nodes_x - 1; ++i) {
      const unsigned base = n_nodes_x * j + i;
      topology.push_back(std::array<unsigned, 2>{base, base + 1});
      topology.push_back(std::array<unsigned, 2>{base, base + n_nodes_x});
      topology.push_back(std::array<unsigned, 2>{base, base + n_nodes_x + 1});
    }
    topology.push_back(std::array<unsigned, 2>{n_nodes_x * (j + 1) - 1,
                                               n_nodes_x * (j + 2) - 1});
  }
  for (unsigned i = 0; i < n_nodes_x - 1; ++i)
    topology.push_back(std::array<unsigned, 2>{
        n_nodes_x * (n_nodes_y - 1) + i, n_nodes_x * (n_nodes_y - 1) + i + 1});

  // Node coords
  std::vector<double> x_unshuffled(n_nodes);
  std::vector<double> y_unshuffled(n_nodes);
  unsigned coord_ind = 0;
  for (unsigned j = 0; j < n_nodes_y; ++j)
    for (unsigned i = 0; i < n_nodes_x; ++i) {
      x_unshuffled[coord_ind] = i;
      y_unshuffled[coord_ind] = j;
      ++coord_ind;
    }

  // Shuffle
  std::mt19937 prng{seed};
  std::vector<unsigned> permutation(n_nodes);
  std::iota(permutation.begin(), permutation.end(), 0u);
  std::shuffle(permutation.begin(), permutation.end(), prng);
  unsigned i = 0;
  for (auto p : permutation) {
    x[p] = x_unshuffled[i];
    y[p] = y_unshuffled[i];
    ++i;
  }
  for (auto &[n1, n2] : topology) {
    n1 = permutation[n1];
    n2 = permutation[n2];
  }
  std::shuffle(topology.begin(), topology.end(), prng);
  return topology;
}

File: perf-ninja-rs/labs/memory_bound/huge_pages_1/src/generate_mesh.rs
// Generate an example mesh describing a 2D truss.
// In this example, the nodes are distributed on a cartesian grid, the topology
// of the mesh is as follows:
/*
 * o-o-o-o-o-o  ^
 * |\|\|\|\|\|  |
 * o-o-o-o-o-o  |  n_nodes_y layers
 * |\|\|\|\|\|  |
 * o-o-o-o-o-o  |
 *
 * ----------->
 * n_nodes_x layers
 *
 */

use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};

// n_nodes_x, n_nodes_y - see sketch above
// x, y - arrays where node coordinates will be written, need to have space
//        allocated for at least (n_nodes_x*n_nodes_y) doubles
// seed - seed for RNG
//
// returns topology (see lib.rs)
pub fn generate_mesh(
    n_nodes_x: u32,
    n_nodes_y: u32,
    x: &mut [f64],
    y: &mut [f64],
    seed: u64,
) -> Vec<[u32; 2]> {
    let n_nodes = n_nodes_x * n_nodes_y;

    // Topology
    let mut topology: Vec<[u32; 2]> = Vec::with_capacity(
        ((n_nodes_x - 1) * (n_nodes_y - 1) * 3 + (n_nodes_x - 1) + (n_nodes_y - 1)) as usize,
    );
    for j in 0..(n_nodes_y - 1) {
        for i in 0..(n_nodes_x - 1) {
            let base = n_nodes_x * j + i;
            topology.push([base, base + 1]);
            topology.push([base, base + n_nodes_x]);
            topology.push([base, base + n_nodes_x + 1]);
        }
        topology.push([n_nodes_x * (j + 1) - 1, n_nodes_x * (j + 2) - 1]);
    }
    for i in 0..(n_nodes_x - 1) {
        topology.push([
            n_nodes_x * (n_nodes_y - 1) + i,
            n_nodes_x * (n_nodes_y - 1) + i + 1,
        ]);
    }

    // Node coords
    let mut x_unshuffled = vec![0.0; n_nodes as usize];
    let mut y_unshuffled = vec![0.0; n_nodes as usize];
    let mut coord_ind: usize = 0;
    for j in 0..n_nodes_y {
        for i in 0..n_nodes_x {
            x_unshuffled[coord_ind] = i as f64;
            y_unshuffled[coord_ind] = j as f64;
            coord_ind += 1;
        }
    }

    // Shuffle
    let mut prng = StdRng::seed_from_u64(seed);
    let mut permutation: Vec<f64> = (0..n_nodes).map(|x| x as f64).collect();
    permutation.shuffle(&mut prng);
    x[..n_nodes as usize].copy_from_slice(&x_unshuffled[..n_nodes as usize]);
    y[..n_nodes as usize].copy_from_slice(&y_unshuffled[..n_nodes as usize]);

    for [n1, n2] in topology.iter_mut() {
        *n1 = permutation[*n1 as usize] as u32;
        *n2 = permutation[*n2 as usize] as u32;
    }
    topology.shuffle(&mut prng);

    topology
}
